src/query/storage/fanout/storage.go

```diff
@@ -81,18 +81,58 @@ func (s *fanoutStorage) FetchBlocks(
 	query *storage.FetchQuery,
 	options *storage.FetchOptions,
 ) (block.Result, error) {
-	stores := filterStores(s.stores, s.writeFilter, query)
-	blockResult := block.Result{}
+	stores := filterStores(s.stores, s.fetchFilter, query)
+	// Optimization for the single store case
+	if len(stores) == 1 {
+		return stores[0].FetchBlocks(ctx, query, options)
+	}
+
+	// TODO(arnikola): use a genny map here instead, also execute in parallel.
+	blockResult := make(map[string]block.Block, 10)
 	for _, store := range stores {
 		result, err := store.FetchBlocks(ctx, query, options)
 		if err != nil {
 			return block.Result{}, err
 		}
 
-		blockResult.Blocks = append(blockResult.Blocks, result.Blocks...)
+		for _, bl := range result.Blocks {
+			it, err := bl.SeriesIter()
+			if err != nil {
+				return block.Result{}, err
+			}
+
+			key := it.Meta().Bounds.String()
+			if foundBlock, found := blockResult[key]; found {
+				// this block exists. Check to see if it's already an appendable block.
+				if acc, ok := foundBlock.(block.AccumulatorBlock); ok {
+					// already an accumulator block, add current block.
+					if err := acc.AddBlock(bl); err != nil {
+						return block.Result{}, err
+					}
+				} else {
+					acc := block.NewContainerBlock(2)
```

Comment:
nit: Make this size a const at top of file, and/or a default.

Reply:
Refactored a bit